app/assets/javascripts/app/utils.js

```diff
@@ -6,23 +6,55 @@ const dismiss = '[data-dismiss="true"]';
 $(document).on('click', dismiss, (e) => { $(e.target).parent().remove(); });
 
 
-// Safari & IE 8/9 do not support client side handling of `required` attribute on
-// form inputs; this adds basic messaging and styling fallback for these browsers
-// "page:load" event needed because of turbolinks, which overrides normal loading process
+// Safari & IE 8/9 do not support client side validation of form inputs; this adds basic messaging
+// and styling fallback for these browsers "page:load" event needed because of turbolinks, which
+// overrides normal loading process
 $(document).on('ready page:load', () => {
-  const message = '<div class="error-notify alert-danger p1 mt1 mb2">Please fill in all required' +
-    ' fields.</div>';
+  // Detect if HTML5 form validation is available
+  function hasHtml5Validation () {
+    return typeof document.createElement('input').checkValidity === 'function';
+  }
+  // If HTML5 form validation is not supported, or Safari is being used:
+  if (!hasHtml5Validation ||
+    (navigator.userAgent.indexOf('Safari') !== -1
+      && navigator.userAgent.indexOf('Chrome') === -1)) {
+    $('form').on('submit', function(e) {
+      if (!this.checkValidity()) {
+        const $form = $(this);
+        const $required = $form.find('[required]').filter(function() { return this.value === ''; });
+        const $pattern = $form.find('[pattern]').filter(function() { return this.value; });
+        const message = '<div class="error-notify alert-danger p1 mt1 mb2">' +
+          'Please fill in all required fields using the requested format.</div>';
 
-  $('form').on('submit', function(e) {
-    const $form = $(this);
-    const $fields = $form.find('[required]').filter(function() { return this.value === ''; });
+        // Clean up any old error messages
+        $('input', this).removeClass('border-red');
+        $('.error-notify').remove();
 
-    if ($fields.length) {
-      e.preventDefault();
-      if (!$('.error-notify').length) {
-        $form.prepend(message);
+        // Highlight any required fields without a value
+        if ($required.length) {
+          e.preventDefault();
+          $required.each(function() { $(this).addClass('border-red'); });
+        }
+
+        // Highlight any fields that do not match requested pattern
+        let patternValid = true;
```

Comment:
to be a little more clear (and be consistent with indicator variables above), let's prepend `has` or `is` to the name (i.e., const `hasInvalidPattern` or something)